### ISSUE TITLE
wmiexec doesn't work for me

### ISSUE THREAD

**Issue by sshock** [CoreSecurity/impacket#72]
I tried several targets and wmiexec never works for me.  Here are some sample outputs with -debug flag.

On some targets I get this:

    Impacket v0.9.14-dev - Copyright 2002-2015 Core Security Technologies
    Password:
    [*] SMBv2.1 dialect used
    [-] rpc_s_access_denied

On some targets I get this:

    Impacket v0.9.14-dev - Copyright 2002-2015 Core Security Technologies
    [*] SMBv2.1 dialect used
    [-] DCOM SessionError: unknown error code: 0x800706ba

On some targets I get something like this, and then wmiexec.py hangs and I have to kill it:

    Impacket v0.9.14-dev - Copyright 2002-2015 Core Security Technologies
    [*] SMBv2.1 dialect used
    [+] StringBinding: \\\\MYMACHINE[\\PIPE\\atsvc]
    [+] StringBinding: MYMACHINE[49155]
    [+] StringBinding: 10.10.250.100[49155]
    [+] StringBinding: \\\\MYMACHINE[\\PIPE\\srvsvc]
    [+] StringBinding: \\\\MYMACHINE[\\PIPE\\browser]
    [-] [Errno -2] Name or service not known

**Comment by asolino** [comment]
Please update your fork and run this again, with -debug on. 

Ideally, you should not get `[Errno -2] Name or service not known` anymore.

Do you have access on those machines you get error `0x800706ba`. If so, can you try disabling completely the firewall and try again?

**Comment by sshock** [comment]
Hi Alberto.  I retried the third target machine w/ your latest code and the output is slightly different:

    Impacket v0.9.14-dev - Copyright 2002-2015 Core Security Technologies
    
    [*] SMBv2.1 dialect used
    [+] StringBinding: \\\\MYMACHINE[\\PIPE\\atsvc]
    [+] StringBinding: MYMACHINE[49155]
    [+] StringBinding chosen: ncacn_ip_tcp:MYMACHINE[49155]
    [-] [Errno -2] Name or service not known

When I get a minute I'll try disabling the firewall on this machine, which I do have full control over, and see if that changes anything.

**Comment by asolino** [comment]
Hey.. thanks for your answer.

Can you include the command line you use when executing `wmiexec`?. Just be sure you either use the target's IP address this time.

thanks!

**Comment by sshock** [comment]
Wow, when I try it with the IP address it works!  That is, it solves the problem for the target that was giving me [Errno -2] Name or service not known.

    sshock@sshock:/usr/local/bin/impacket$ ./wmiexec.py -debug -hashes (hidden):(hidden) DOMAIN/user@10.10.250.100 "cmd /c dir c:\\"                         
    Impacket v0.9.14-dev - Copyright 2002-2015 Core Security Technologies
    
    [*] SMBv2.1 dialect used
    [+] StringBinding: \\\\MYMACHINE[\\PIPE\\atsvc]
    [+] StringBinding: MYMACHINE[49155]
    [+] StringBinding: 10.10.250.100[49155]
    [+] StringBinding chosen: ncacn_ip_tcp:10.10.250.100[49155]
     Volume in drive C has no label.
     Volume Serial Number is (hidden)
    
     Directory of c:\
    
    12/20/2010  11:48 AM    <DIR>          folder
    ...(clipped)...

**Comment by sshock** [comment]
I've now found 3 different machines where using IP address instead of hostname fixes the [Errno -2] problem, yes!

Using IP address however does not help with the targets producing the rpc_s_access_denied or the "unknown error code: 0x800706ba".  Those are probably firewall related or something like RPC service not running maybe.

**Comment by sshock** [comment]
FYI, on the target machine that was giving "unknown error code: 0x800706ba", disabling the Windows Firewall fixed the problem.

**Comment by sshock** [comment]
I have some more info on the [Errno -2].  The problem occurs when connecting from my home to targets on my work network (over a vpn).  If I instead connect to them directly from a computer on the work network, it works fine using the hostname.  It only has the problem when connecting from my home computer.

**Comment by asolino** [comment]
Thanks for all the details. Good to know things are moving forward.

1) With regards to the firewall configuration, check this [link](https://msdn.microsoft.com/en-us/library/aa822854(v=vs.85).aspx) to enable the rules needed for WMI access in the firewall while leaving the rest of the restrictions in place.

2) You last comment was useful. May I ask how did you execute the first time it failed? Was it this way? `./wmiexec.py -debug -hashes (hidden):(hidden) DOMAIN/user@MYMACHINE` or `./wmiexec.py -debug -hashes (hidden):(hidden) DOMAIN/user@MYMACHINE.DOMAIN`
There's clearly a DNS resolve issue but I don't understand how the first communication was established (before it failed of course, but there was a communication already in place).

**Comment by sshock** [comment]
Thanks Alberto for the link to firewall configuration.  I am confident that I will be able to use that to fix any computers giving me "unknown error code: 0x800706ba".

I think we are getting to the bottom of [Errno -2].  Technically I wasn't trying either one of the ways you wrote, but I tried them now in addition to what I was trying before and here are the outcomes:

A. `DOMAIN/user@mymachine` => Fails immediately with Errno -2.  I never tried this until just now.
B. `DOMAIN/user@mymachine.DOMAIN` => Fails immediately with Errno -2.  I never tried this until just now.
C. `DOMAIN/user@mymachine.domain.com` => Fails with Errno -2 after some debug messages.  This is what I was doing.

I'm confident that A and B fail because it can't even connect because it can't get the IP address.  Neither "mymachine" nor "mymachine.DOMAIN" will resolve with my DNS from home.  Only mymachine.domain.com will resolve, which is the FQDN of the machine.

Now for the interesting part.  If I add "10.10.250.100 mymachine" to my /etc/hosts then both A and C work!  (B still doesn't work, but I don't care about that one and I would never plan to write it that way).

Based on the debug output, I think impacket is correctly using `mymachine.domain.com` for resolving to get IP address and connect initially, and it is correctly shortening it to `MYMACHINE` in contexts where just a machine name is expected, but then I think maybe it tries to resolve just `mymachine` with DNS and that is where the problem is.

One thing you might want to consider is adding support for a `-I ipaddr` option, to specify the IP address for connecting in cases where the targetName cannot be resolved by DNS.  I know programs like smbclient have this.  Then I could write it like A but pass the -I ipaddr and it would probably work.  I think that would be good to have, but I also think we could make example C above work fine as is, if my conclusion is correct.

**Comment by asolino** [comment]
Hey mate. 

You assessment is correct. I just committed https://github.com/CoreSecurity/impacket/commit/7ac7b28bff5865c47ea833812b9dbd6a65c2c3a4 that will help solving this issue. Could you please `git pull` your repo and test again using `DOMAIN/user@mymachine.domain.com `?

Point taken on `-I ipaddr` option. Sadly it is not straight forward to implement due to the way the library is implemented but certainly possible. Feel free to open an issue for it.

What I see happening in this particular case is the DCOM engine at the target answers back endpoints to connect referencing them with their NetBIOS (e.g. `MYMACHINE`) name or IP Address (e.g. `10.10.250.100`). NetBIOS names usually cannot be resolved unless: 
a. You are in the same network at the target machine and the NetBIOS name is resolved by WINS or similar
b. You specify a DNS suffix, and your DNS can resolve it. 

In an Active Directory environment, usually the DNS Server is the Domain Controller (or is controlled by the DC). For every machine joined to the domain there's an entry with it's NetBIOS name (and maybe some aliases). The trick tho is the DHCP Server usually sends the DNS data specifying to add the domain FQDN suffix into the resolv.conf file. I suspect this is the issue you were experiencing.

Specifying the IP address was working because the library will pick the answers of the target's DCOM engine matching that IP address, hence no resolving is needed.
If you specify names, the library will try to match names when parsing the target's DCOM engine answer, which is important if your using Kerberos authentication.

Long story short, I hope this solves the issue.. if not we'll keep trying ;)

cheers,
beto

**Comment by sshock** [comment]
Thanks.  That fixed the issue!

**Comment by asolino** [comment]
Excellent. Thanks for your help on this Phillip.
